### PR TITLE
Standalone Processor Server (#884 + #974)

### PR DESCRIPTION
**Pull request by MehmedGIT**
This draft PR is supposed to get important aspects from https://github.com/OCR-D/core/pull/884 and adapt to https://github.com/OCR-D/core/pull/974 so it can extend the `ocrd_network` package to provide the standalone Processor Server feature.

I am creating this a bit early so it's easy to follow what is going on with this. ~~Still broken and fixes are needed~~.  ~~A lot more refactoring is needed to follow afterward~~. EDIT: Refactored and on the surface seems to work now.

Feel free to provide feedback.

### CODE REVIEW

**Comment by MehmedGIT** [comment]
@bertsky 

> Looks good already.

Not yet. There is still duplication of code. Mainly in the deployment phase.

> This is still far away from a review, so please restrain to do so yet.

The reason I said that was to keep #1030 small because most of the things you mentioned in your review would have probably changed in the coming commits. For me going through your review for the current time being and answering everything separately in detail is more time-consuming than actually fixing things. Especially, when things are still not optimized and duplication of code is still out there. Also, the logging in `ocrd_network` will change completely after the logging change in the core.

File: ocrd/ocrd/decorators/ocrd_cli_options.py [hunk 1]
```diff
@@ -33,6 +34,8 @@ def cli(mets_url):
         parameter_option,
         parameter_override_option,
         loglevel_option,
+        option('--agent_type', type=click.STRING),
```
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd/ocrd/decorators/ocrd_cli_options.py:37]: I like the suggestion. This is much cleaner, yes. The only difference is that I will still stick to `agent_type` instead of `type_` internally. Not a big fan of hacky workarounds that are hard to read and follow than more descriptive naming conventions. The core base is already hard to follow (for external developers) due to the same reason... In case you missed it, check the discussion [here](https://github.com/OCR-D/core/pull/974#discussion_r1145327980) as well.

File: ocrd/ocrd/decorators/ocrd_cli_options.py [hunk 2]
```diff
@@ -33,6 +34,8 @@ def cli(mets_url):
...
+        option('--agent_address', type=click.STRING),
```
[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd/ocrd/decorators/ocrd_cli_options.py:38]: > So in short, **`this is still far away from a review, so please restrain to do so yet`**.

I hear you. The least thing I want is put extra work on your shoulders. The review was not meant that way though. I just tried to give some early commentary, in a mix of
- problems already visible, that will likely fall on your feet (e.g. the processor decorator/constructor contexts)
- small ideas you might find useful
- concerns (based on my limited understanding) which might indicate potential problems going forward

You don't need to address everything. But I'll keep quiet until you release the draft.

> The `client`, in the way you imagined it, is currently the `Processing Server` which redirects coming requests to deployed `Processor Servers` (i.e., processing worker as a standalone server - but both concepts are separated to avoid confusion).

I understand that. But the Processing Server **also** has value as a replacement for the processor CLI. It's complementary. IPs will have different models for different tasks – a full-blown Web-API backed Workflow Server with Processing Server and queues, a Processing Server without queues only Processor Servers, and finally just some scripts or custom tools with the simple CLI interface to processors. To make the latter at least more easily installable and scalable, a backwards compatible CLI interface to the Processor Servers makes sense, and would allow splitting up ocrd_all (now 36 GB!) into single module containers. The increasing problems getting ocrd_all workable in the current (fat container – few venvs) form is part of the urgency.

> See, even you have mistaken the usage between the `processor server` with the `processing worker`. The `processor server` does not take `--queue Q` as a parameter.

Yes, sorry, that was sloppy writing. Cut the `--queue Q` in my example.

> I think this is not something we would like to have in that form, do we? Correct me if I am wrong, but why would a user need to install the processor-executable on their machine only to be able to send a request to the `Processing Server` or `Processor Server` (which may not be running on their machine)? 

It wouldn't be an installation of the actual processor, just of core. Perhaps ocrd_all's recipes could install a client tool named after each processor to be installed, and the client tool looks up its name to determine where to connect. (So much for my idea.)

> How about just:
> 
> ```shell
> ocrd network client --address localhost:1000 -m path/to/mets.xml -I OCR-D-SEG -O OCR-D-OCR -P checkpoint_dir default
> ```

Yes, that would do it, too.

> Or:
> 
> ```shell
> ocrd network client --address localhost:1000 ocrd-calamari-recognize -m path/to/mets.xml -I OCR-D-SEG -O OCR-D-OCR -P checkpoint_dir default
> ```
> 
> Where `--address` is `Processing Server` address and the client doesn't have to bother knowing the exact address of the specific `Processor Server`. The `Processing Server` is a proxy in that case.

Yes, that's even better! So using the Processing Server as a proxy and deployer does have benefits even in this scenario.

> Of course, with proper env variables, we can also drop the `--address` option and resolve it internally.

Yes, I thought so, too. Something in the environment that will tell both the deployer and the client where everything can be found. Could also be the deployer config file, though – perhaps even for the client?

> In the end, these two will be equal:
> 
> ```shell
> ocrd network client ocrd-* -m path/to/mets.xml -I OCR-D-* -O OCR-D-* -P ...
> ocrd-* -m path/to/mets.xml -I OCR-D-* -O OCR-D-* -P ...
> ```

Yes, that's the idea. So our ocrd_all can just provide simple delegator scripts, and everything can be made to look like CLI installation, as before. (Of course, the module containers will have to be allowed to run as Processor Workers if required.)

> Isn't the `ocrd_cli_wrap_processor` method already cluttered enough when a processor can be a `worker` and a `server`, so we need a `client` on top with a different `--address` option as well? While having `worker` and `server` makes sense, since these cannot be called without having the executable in hand, that's not the case for the `client`.

If we do introduce a `network` command group, then – at least for the user – I don't think this is clutter. On the implementation side I don't see the problem either – the decorator just needs another conditional.

File: ocrd_network/ocrd_network/processor_server.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            end_time=end_time,
+            exec_time=f'{exec_duration} ms'
+        )
+        return job.to_job_output()
+
+    def get_ocrd_tool(self):
+        if self.ocrd_tool:
+            return self.ocrd_tool
+        if self.ProcessorClass:
+            ocrd_tool = self.ProcessorClass(workspace=None, version=True).ocrd_tool
```
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd_network/ocrd_network/processor_server.py]: You're right. I just wanted to find a cleaner and more direct way to get the `ocrd_tool` and the `version` from the `ProcessorClass` directly. It was working on the surface so I did not dig into the details. Have already reverted that back to reading a stdout and parsing that instead as a temporal solution.

File: ocrd_network/ocrd_network/processor_server.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            ).stdout
+        return version_str
+
+    def run_server(self, host, port, access_log=False):
+        # TODO: Provide more flexibility for configuring file logging (i.e. via ENV variables)
+        file_handler = logging.FileHandler(f'/tmp/server_{self.processor_name}_{port}_{getpid()}.log', mode='a')
+        logging_format = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
+        file_handler.setFormatter(logging.Formatter(logging_format))
+        file_handler.setLevel(logging.DEBUG)
+        self.log.addHandler(file_handler)
```
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd_network/ocrd_network/processor_server.py:238]: Yes, this will change with the changes to the core logging.

File: ocrd_network/ocrd_network/processor_server.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+# TODO: Check this again when the logging is refactored
+tf_disable_interactive_logs()
+
+
+class ProcessorServer(FastAPI):
+    def __init__(self, mongodb_addr: str, processor_name: str = "", processor_class=None):
+        if not (processor_name or processor_class):
+            raise ValueError('Either "processor_name" or "processor_class" must be provided')
+        self.log = getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd_network/ocrd_network/processor_server.py:45]: Why though? Maybe core needs to change to `__name__`? It's a cleaner approach and always resolves to the correct `module_name.file_name` instead of manually naming the logger. So in that specific case resolves to `ocrd_network.processor_server`.

File: ocrd/ocrd/decorators/__init__.py [hunk 6]
```diff
@@ -56,96 +55,127 @@ def ocrd_cli_wrap_processor(
...
+                # TODO: Better validate that inside the ProcessorServer itself
+                host, port = agent_address.split(':')
+                processor_server = ProcessorServer(
+                    mongodb_addr=database,
+                    processor_name=processor.ocrd_tool['executable'],
+                    processor_class=ProcessorClass,
+                )
+                processor_server.run_server(host=host, port=int(port))
+            except Exception as e:
+                sys.exit(f"Processor server has failed with error: {e}")
```
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd/ocrd/decorators/__init__.py]: Resolved.

File: ocrd/ocrd/decorators/__init__.py [hunk 7]
```diff
@@ -56,96 +55,127 @@ def ocrd_cli_wrap_processor(
...
+            pstats.Stats(pr, stream=s).sort_stats("cumulative").print_stats()
+            print(s.getvalue())
+        atexit.register(exit)
+    run_processor(processorClass, mets_url=mets, workspace=workspace, **kwargs)
+
+
+def check_and_run_network_agent(ProcessorClass, agent_type: str, agent_address: str, database: str, queue: str):
+    if not agent_type and (agent_address or database or queue):
+        raise ValueError("Options '--database', '--queue', and 'agent_address' are valid only with '--agent_type'")
+    if agent_type:
```
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd/ocrd/decorators/__init__.py]: Resolved.

File: ocrd_network/ocrd_network/process_helpers.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import json
+from typing import List
+
+from ocrd import Resolver
+from ocrd.processor.helpers import run_cli, run_processor
+
+
+# A wrapper for run_processor() and run_cli()
+def run_single_execution(
```
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd_network/ocrd_network/process_helpers.py:9]: I would stick to `run_single_execution` for the current time being since `process` is too wide and does not hint at what it refers to directly. Also, there are `process_message` and `processor_job_task`. As agreed before, the naming conventions should be discussed at the later stages when everything is working and ready for merge/release.

File: ocrd_network/ocrd_network/processing_worker.py [hunk 9]
```diff
@@ -83,7 +72,7 @@ def __init__(self, rabbitmq_addr, mongodb_addr, processor_name, ocrd_tool: dict,
         self.processor_name = processor_name
         # The processor class to be used to instantiate the processor
         # Think of this as a func pointer to the constructor of the respective OCR-D processor
-        self.processor_class = processor_class
+        self.ProcessorClass = processor_class
```
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd_network/ocrd_network/processing_worker.py:75]: Purely for readability purposes through the network module - to follow the convention of naming classes.

File: ocrd_network/ocrd_network/processor_server.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd_network/ocrd_network/processor_server.py:45]: Ok, you're right. We started with some conventions, trying to get our logs to look nice (with or without debug level) for core, processor modules and bashlib processors. Perhaps at least for core it's time to walk the Python path again. (Using `__name__` in most places instead of fixed strings, no more `ocrd_cli_wrap_processor` or `ocrd.process.profile`, `ocrd.dummy`.) Still, one difficulty there is to hide details of our implementation (like processor.helpers instead of just processor, or `ocrd_network.server_utils` instead of `ocrd_network.processing_worker` etc).

[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd_network/ocrd_network/processor_server.py:45]: > Still, one difficulty there is to hide details of our implementation

That's problematic indeed.

File: ocrd_network/ocrd_network/processor_server.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+                run_single_execution,
+                ProcessorClass=self.ProcessorClass,
+                executable=self.processor_name,
+                abs_path_to_mets=job.path_to_mets,
+                input_file_grps=job.input_file_grps,
+                output_file_grps=job.output_file_grps,
+                page_id=job.page_id,
+                parameters=job.parameters
+            )
+        except Exception as error:
```
[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd_network/ocrd_network/processor_server.py:180]: https://github.com/OCR-D/core/blob/93482d39e50ded9046cae09e0bc48969ab8d28e4/ocrd_network/ocrd_network/processor_server.py#L143-L150

Yes, external queues are not an option. We definitely need to run within the same process here, otherwise instance caching would not work. @tdoan2010's design choice was to not allow concurrency in these servers (multiple workers via multithreading or multiprocessing backed by shared memory, as in my [Processing Server](https://github.com/OCR-D/core/pull/652)), but instead require running multiple instances if necessary.

But to address FastAPI's problem with error handling (follow-up tasks getting lost) and oversubscription (no requests getting denied), couldn't we also just make the Processor Server a simple blocking API?

[the remark this answers is not in the document]
Comment (tdoan2010) [reply ocrd_network/ocrd_network/processor_server.py:180]: > couldn't we also just make the Processor Server a simple blocking API

Sure, that will simplify a lot of things. But we need to make sure that the `timeout` is set properly then. This can be a parameter set by users when they start up a server.

File: ocrd_network/ocrd_network/processor_server.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+        return version_str
+
+    def run_server(self, host, port, access_log=False):
+        # TODO: Provide more flexibility for configuring file logging (i.e. via ENV variables)
+        file_handler = logging.FileHandler(f'/tmp/server_{self.processor_name}_{port}_{getpid()}.log', mode='a')
+        logging_format = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
+        file_handler.setFormatter(logging.Formatter(logging_format))
+        file_handler.setLevel(logging.DEBUG)
+        self.log.addHandler(file_handler)
+        uvicorn.run(self, host=host, port=port, access_log=access_log)
```
[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd_network/ocrd_network/processor_server.py:239]: I understand, uvicorn (as does gunicorn) does not make it easy to deploy the server programmatically. But we should not dismiss the multiprocessing/worker model outright just because of that.

In https://github.com/OCR-D/core/pull/652, I also [used the CLI to create an app instance with runtime parameters](https://github.com/bertsky/core/blob/d98daa80695bd4bd5b0f23d4435a3abb0d524f54/ocrd/ocrd/decorators/__init__.py#L51-L67), and had my server simply [inherit from gunicorn.app.base.BaseApplication](https://github.com/bertsky/core/blob/d98daa80695bd4bd5b0f23d4435a3abb0d524f54/ocrd/ocrd/server.py#L18-L69) so I could run it – including the preforking multi-worker paradigm.

Here, since we also have to accommodate FastAPI, we would have to do some additional steps:
- use the [same inheritance pattern](https://docs.gunicorn.org/en/stable/custom.html#custom-application) for the ProcessingServer here
- tell [gunicorn to use uvicorn as worker class](https://www.uvicorn.org/#running-with-gunicorn)
- instead of the cmdline string, pass [`worker_class` as runtime option](https://github.com/benoitc/gunicorn/blob/792edf6d9aabcbfb84e76be1d722ac49c32dc027/gunicorn/config.py#L656-L657), pointing to the imported [`uvicorn.workers.UvicornWorker`](https://github.com/encode/uvicorn/blob/16d5101433bbfacf18e6ce9d21041164995b8bbb/uvicorn/workers.py#L14)

File: ocrd_network/ocrd_network/process_helpers.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd_network/ocrd_network/process_helpers.py:9]: > In essence, what this does is delegate to ocrd.Processor.process. So why not just call it ocrd_network.process_helpers.process (or even better ocrd.processor.process)?

Because I wanted a clear differentiation between `ocrd.processor.process` and `ocrd_network.process_helpers.process`. 

> After all, process is what a processor should do.

Yes, but we have Processing Worker and Processor Server context, not Processor. 

> I agree there's a clash with `process_message` (as imperative of OcrdProcessingMessage), but I would rather change the latter (`consume_message` would seem a better fit) than block the former interpretation. 

There is already `on_consumed_message` which decodes `bytes` to `OcrdProcessingMessage`, handles RabbitMQ stuff, and passes the `OcrdProcessingMessage` to `process_message`. The `process_message` then, processes the message by resolving data inside the message.

> To me, what also speaks against `run_single_execution` is that `run` is redundant with `execute`.

Agree, my bad. `invoke_execution`/`invoke_processing` or something along these lines would have been better. Since I don't see this potentially conflicting anywhere in the future, I have refactored `run_single_execution` to `process`.

> IMHO `ocrd_network.ProcessorServer.processor_job_task` should be an imperative verb, too. And it's also redundant btw (`job` vs. `task`). How about `consume_processor_job` or `do_processor_job`, or (replacing the term processor job througout) `create_task` + `run_task`?

Agree. In my mind, conceptually, the job was referring to the job model and I made the mistake to put `job` and `task` together, which is bad. I don't think `consume` fits here, nothing is being consumed. 

In conclusion, I have refactored: 
`run_single_execution` -> `invoke_processor`
`create_processor_job_task` -> `create_processor_task` 
`processor_job_task` -> `run_processor_task`

Thanks for your input, this looks better now!

File: ocrd_network/ocrd_network/processing_worker.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd_network/ocrd_network/processing_worker.py:75]: Sure, but it's still an attribute referencing a class. I don't mind going back to `processor_class` since there is `_class` in any way. `ProcessorClass` -> `processor_class`

File: ocrd/ocrd/decorators/ocrd_cli_options.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd/ocrd/decorators/ocrd_cli_options.py]: Since it's not obvious how to implement `ocrd-* network ...` without changing separate processor implementations should we resolve this? I have already implemented the initial skeleton for the `ocrd network client ...` and the single processing request which seems to work. However, I think the discussion for that should continue under #1032. I will update the current CLI status there today or tomorrow.

File: ocrd_network/ocrd_network/processor_server.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd_network/ocrd_network/processor_server.py]: I am resolving this. Should this continue in a separate discussion?

> Getting the .ocrd_tool from non-processing context is a good idea though. We could work with show_version=True (instead of version=True), which only has the side effect of printing the version to stdout (which can be intercepted or allowed).
Another option is to use dump_json=True, but not to access the .ocrd_tool member afterwards, but directly for its stdout.
Or we implement a new, tailored non-processing context in the constructor, say get_json=True, which does a return instead of a print – but that might not be allowed with Python.

File: ocrd_network/ocrd_network/processor_server.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply ocrd_network/ocrd_network/processor_server.py:239]: > I understand, uvicorn (as does gunicorn) does not make it easy to deploy the server programmatically. But we should not dismiss the multiprocessing/worker model outright just because of that.

Of course not! Let's have a discussion and a clear vision of how we want to implement that so that all of us agree on common ground. @tdoan2010, please join the discussion.

**Comment by MehmedGIT** [comment]
Feel free to provide more extensive feedback now. Things that can be changed immediately, will be changed. Things that need further discussions, should wait till all of us reach an agreement on how things should be implemented.

File: ocrd_network/ocrd_network/processor_server.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd_network/ocrd_network/processor_server.py]: Not necessary – you already resolved all that with your changes.

File: ocrd/ocrd/decorators/ocrd_cli_options.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd/ocrd/decorators/ocrd_cli_options.py]: Yes, I agree we should resolve. Fighting for CLI grouping in our processor decorator is not worth the effort, if possible at all. And I think we already agreed that since we have a custom help generator which already can group the various options, true command grouping is not even necessary.

File: ocrd_validators/ocrd_validators/processing_server_config.schema.yml [hunk 12]
```diff
@@ -58,6 +58,7 @@ properties:
         - address
         - username
         - workers
```
Comment (joschrew) [inline ocrd_validators/ocrd_validators/processing_server_config.schema.yml:60]: I think it would be good to allow only workers or only servers. For that I suggest replacing the following:
```
        - workers
        - servers
      oneOf:
        - required:
            - password
        - required:
            - path_to_privkey
```
with this
```
      allOf:
        - oneOf:
          - required:
              - password
          - required:
              - path_to_privkey
        - oneOf:
          - required:
              - workers
          - required:
              - servers
```

File: ocrd_network/ocrd_network/runtime_data.py [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+                self.needs_ssh = True
+            if not self.needs_docker and deploy_type == DeployType.DOCKER:
+                self.needs_docker = True
+            for _ in range(count):
+                self.data_workers.append(DataProcessingWorker(self.address, deploy_type, name))
+
+        for server in config['servers']:
+            name = server['name']
+            port = server['port']
+            deploy_type = DeployType.DOCKER if server['deploy_type'] == 'docker' else DeployType.NATIVE
```
Comment (joschrew) [inline ocrd_network/ocrd_network/runtime_data.py:52]: This is needed to allow only workers or only servers (and small bugfix if deploy_type not provided as allowed in schema)
```suggestion
        for worker in config.get('workers', []):
            name = worker['name']
            count = worker['number_of_instance']
            deploy_type = DeployType.DOCKER if worker['deploy_type'] == 'docker' else DeployType.NATIVE
            if not self.needs_ssh and deploy_type == DeployType.NATIVE:
                self.needs_ssh = True
            if not self.needs_docker and deploy_type == DeployType.DOCKER:
                self.needs_docker = True
            for _ in range(count):
                self.data_workers.append(DataProcessingWorker(self.address, deploy_type, name))

        for server in config.get('servers', []):
            name = server['name']
            port = server['port']
            deploy_type = DeployType.DOCKER if server.get('deploy_type', None) == 'docker' else DeployType.NATIVE
```

File: ocrd_validators/ocrd_validators/processing_server_config.schema.yml (hunk 12 above)
Comment (MehmedGIT) [reply ocrd_validators/ocrd_validators/processing_server_config.schema.yml:60]: Agree. As it is currently, it restricts starting only workers or only servers. However, I think it should be `anyOf` instead of `oneOf`. I would still keep it flexible and allow deploying workers and servers simultaneously.

Comment (joschrew) [reply ocrd_validators/ocrd_validators/processing_server_config.schema.yml:60]: Yes, right, `anyOf` is better

File: ocrd_network/ocrd_network/runtime_data.py (hunk 13 above)
Comment (joschrew) [reply ocrd_network/ocrd_network/runtime_data.py:52]: When I tested I provided one server without `deploy_type` specified. IIRC the spec allows that (because native is the default) and it failed with an exception. And my line 52 edits fixed that exception.

File: ocrd_validators/ocrd_validators/processing_server_config.schema.yml (hunk 12 above)
Comment (MehmedGIT) [reply ocrd_validators/ocrd_validators/processing_server_config.schema.yml]: Resolved. Now it's possible to start only workers or standalone servers or both.

File: ocrd_network/ocrd_network/processor_server.py [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+        execution_failed = False
+        start_time = datetime.now()
+        await db_update_processing_job(
+            job_id=job_id,
+            state=StateEnum.running,
+            start_time=start_time
+        )
+        try:
+            background_tasks.add_task(
```
[the remark this answers is not in the document]
Comment (tdoan2010) [reply ocrd_network/ocrd_network/processor_server.py]: Already solved through our discussion. We use the blocking mechanism, so `background_tasks` is not needed anymore.

**Comment by MehmedGIT** [comment]
@kba, the blocking API for the Processor Servers and non-blocking requests to the Processor Servers from Processing Server have been implemented. This PR should now be good to be merged.

File: ocrd_network/ocrd_network/processor_server.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd_network/ocrd_network/processor_server.py:161]: @MehmedGIT you now [implemented](https://github.com/OCR-D/core/pull/1030/commits/4b5d48ea680238059e4e71d39a63ed63f364d92e) this as a client-side timeout (with a dummy for the number of pages). I could live with that, but we must not forget to
- catch the additional ReadTimeout
- make this configurable (in the Processing Server now; and in the Processor Server client CLI)
- use the actual number of pages (by resolving the job's `page_id` param against its `workspace_id` or `path_to_mets` param, and then counting the remaining number of physical pages on the workspace)

In contrast, if we do place this on the side of the Processor Server as @tdoan2010 proposed (I guess that would be around the [`invoke_processor` call](https://github.com/OCR-D/core/blob/4b5d48ea680238059e4e71d39a63ed63f364d92e/ocrd_network/ocrd_network/processor_server.py#L143)), we would have the chance of passing the timeout param down (over API or CLI) to the actual Processor instance. (Which the Processing Worker could also benefit from BTW.) So in that scenario, on the caller side, we would get an additional error case (timeout).

[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd_network/ocrd_network/processor_server.py:161]: > I thought dummy timeout for now is still better than no timeout at all.

Yes, it's good to have it in the code already. Perhaps I should have created an issue right away instead of prolonging the discussion here.

> > catch the additional ReadTimeout
> 
> Could you elaborate on that a bit more? 

If you set a [`timeout`](https://www.python-httpx.org/advanced/#timeout-configuration) on httpx.Client or AsyncClient, then it can raise TimeoutException (or perhaps one of its [subclasses](https://www.python-httpx.org/exceptions/), which I assume would have to be ReadTimeout here, because the connection itself already has been established and we are running out of time while trying to read the response).

> I am just not sure what is best for now. Providing an extra field for request timeout or for amount of pages, so the timeout could be then calculated.

I am convinced generic timeouts make no sense for us, so calculation relative to number of pages is crucial.

> When there is a page range, sure. What about the cases when no `page_id` is provided?

Even if there is a page range parameter value (list or regexp), it must be resolved against the actual pages, which yields the true list (that could be shorter than the input list or empty). So regardless, one needs access to the workspace instance (deeper inside the call tree). Perhaps the parameter validator could also give us that number already.

**Comment by kba** [comment]
AFAICT the only open discussion is the one around timeouts - shall I move this to a separate issue?

Otherwise, I'll merge after our call later, many thanks!

**Comment by bertsky** [comment]
Now or never!

File: ocrd_network/ocrd_network/processor_server.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kba) [reply ocrd_network/ocrd_network/processor_server.py:161]: OK, for now we have the hardcoded 200 pages and a TODO, let's not let that hinder merging now and revisiting for proper page list information from the workspace later.

**Review by kba — APPROVED** [review]
